### ISSUE TITLE
POST requests to start endpoint fails with ContentLength error

### ISSUE THREAD

**Issue by FrenchBen** [moby/moby#33501]
**Description**

Making an API request to start a container with an empty JSON string in the body will return the following error:
`{"message":"starting container with non-empty request body was deprecated since v1.10 and removed in v1.12"}`

This functionality worked in all previous versions of docker but is now broken in 17.06.0-rc1

**Steps to reproduce the issue:**
1. Create a container
2. Send a request to start the container via the API and an empty JSON string `{}`
3. Get the above error

```
$ docker create --name my-test alpine top
$ curl -vvv --unix-socket /var/run/docker.sock -X POST -H 'Content-Type: application/json' -d '{}' http://1.29/containers/my-test/start
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying /var/run/docker.sock...
...
> POST /containers/my-test/start HTTP/1.1
> Host: 1.29
> User-Agent: curl/7.51.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 2
>
* upload completely sent off: 2 out of 2 bytes
< HTTP/1.1 400 Bad Request
< Api-Version: 1.30
< Content-Length: 109
< Content-Type: application/json
< Date: Fri, 02 Jun 2017 21:57:20 GMT
< Docker-Experimental: true
< Ostype: linux
< Server: Docker/17.06.0-ce-rc1 (linux)
<
{"message":"starting container with non-empty request body was deprecated since v1.10 and removed in v1.12"}
* Curl_http_done: called premature == 0
* Connection #0 to host 1.29 left intact
```
To make it work:
```
$ curl --unix-socket /var/run/docker.sock -X POST -H 'Content-Type: application/json' -d '' http://1.29/containers/my-test/start
```

**Describe the results you received:**
`{"message":"starting container with non-empty request body was deprecated since v1.10 and removed in v1.12"}`

**Describe the results you expected:**
For the container to start

**Additional information you deem important (e.g. issue happens only occasionally):**
https://github.com/moby/moby/blob/master/api/server/router/container/container_routes.go#L144

**Output of `docker version`:**

```                                                                                                                                          
Client:
 Version:      17.06.0-ce-rc1
 API version:  1.30
 Go version:   go1.8.1
 Git commit:   7f8486a
 Built:        Wed May 31 02:56:01 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.06.0-ce-rc1
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.1
 Git commit:   7f8486a
 Built:        Wed May 31 03:00:14 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Version 17.06.0-rc1-ce-mac13 (18169)

**Comment by thaJeztah** [comment]
@FrenchBen probably not related, but your url seems incorrect (the API version is ignored, because it's missing a hostname; should be `http://localhost/1.29/containers/mytest/start`

**Comment by thaJeztah** [comment]
Wondering if this is a proxy issue;

Running on the Mac;

```bash
curl -vvv --unix-socket /var/run/docker.sock -X POST -H 'Content-Type: application/json' -d '{}' http://localhost/v1.29/containers/foo/start
Note: Unnecessary use of -X or --request, POST is already inferred.

*   Trying /var/run/docker.sock...
* Connected to localhost (/Users/sebastiaan/Library/Containers/com.dock) port 80 (#0)
> POST /v1.29/containers/foo/start HTTP/1.1
> Host: localhost
> User-Agent: curl/7.51.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 2
>
* upload completely sent off: 2 out of 2 bytes
< HTTP/1.1 400 Bad Request
< Api-Version: 1.30
< Content-Length: 109
< Content-Type: application/json
< Date: Sat, 03 Jun 2017 00:45:51 GMT
< Docker-Experimental: true
< Ostype: linux
< Server: Docker/17.06.0-ce-rc1 (linux)
<
{"message":"starting container with non-empty request body was deprecated since v1.10 and removed in v1.12"}
* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact
```

Running the same from within a container (docker.sock mounted);

```bash
docker run \
  -v /var/run/docker.sock:/var/run/docker.sock \
  --rm \
  alpine sh -c 'apk add --no-cache curl && curl -vvv --unix-socket /var/run/docker.sock -X POST -H \'Content-Type: application/json\' -d \'{}\' http://localhost/v1.29/containers/foo/start'

fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/community/x86_64/APKINDEX.tar.gz
(1/4) Installing ca-certificates (20161130-r1)
(2/4) Installing libssh2 (1.7.0-r2)
(3/4) Installing libcurl (7.52.1-r3)
(4/4) Installing curl (7.52.1-r3)
Executing busybox-1.25.1-r0.trigger
Executing ca-certificates-20161130-r1.trigger
OK: 5 MiB in 15 packages
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying /var/run/docker.sock...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to localhost (/var/run/docker.sock) port 80 (#0)
> POST /v1.29/containers/foo/start HTTP/1.1
> Host: localhost
> User-Agent: curl/7.52.1
> Accept: */*
> Content-Type: application/json
> Content-Length: 2
>
} [2 bytes data]
* upload completely sent off: 2 out of 2 bytes
< HTTP/1.1 204 No Content
< Api-Version: 1.30
< Docker-Experimental: true
< Ostype: linux
< Server: Docker/17.06.0-ce-rc1 (linux)
< Date: Sat, 03 Jun 2017 00:48:26 GMT
<
* Curl_http_done: called premature == 0
100     2    0     0  100     2      0      5 --:--:-- --:--:-- --:--:--     5
* Connection #0 to host localhost left intact
```

**Comment by FrenchBen** [comment]
Ping @djs55

**Comment by KostyaSha** [comment]
Hello from @docker-java Probably proxy is screwing content-length https://github.com/docker-java/docker-java/pull/856#issuecomment-306254980

**Comment by djs55** [comment]
After enabling verbose debug logs in the proxy, I see this:
```
2017/06/06 10:37:16 proxy underlying <- "POST /v1.29/containers/foo/start HTTP/1.1\r\nHost: localhost\r\nUser-Agent: curl/7.51.0\r\nTransfer-Encoding: chunked\r\nAccept: */*\r\nContent-Type: application/json\r\n\r\n0\r\n\r\n"
2017/06/06 10:37:16 proxy underlying -> "HTTP/1.1 400 Bad Request\r\nApi-Version: 1.30\r\nContent-Type: application/json\r\nDocker-Experimental: true\r\nOstype: linux\r\nServer: Docker/17.06.0-ce-rc1 (linux)\r\nDate: Tue, 06 Jun 2017 09:37:16 GMT\r\nContent-Length: 109\r\n\r\n{\"message\":\"starting container with non-empty request body was deprecated since v1.10 and removed in v1.12\"}\n"
```
so `curl` claims to send a non-chunked `Content-Length: 2` while the proxy forwards a `Transfer-Encoding: chunked` block of length 0. Somehow the body has gone missing.

(As an aside, I'm a bit confused about the error "starting container with non-empty request body was deprecated"-- the proxy seems to have sent an empty body, I don't know why it's complaining about a non-empty body)

**Comment by KostyaSha** [comment]
Because docker checks for content length https://github.com/moby/moby/blob/master/api/server/router/container/container_routes.go#L139-L147

**Comment by thaJeztah** [comment]
This is fixed in the 17.06.0-rc2 release of Docker for Mac
